### PR TITLE
V2: Fix small hpopt bugs and add example notebook

### PR DESCRIPTION
**Pull request by hwpang**
## Description
Add an example hpopt notebook. Happy to take suggestion on the notebook.
Found some small bugs related to input type and saving hpopt results while working on the notebook. Fixed those.

## Example / Current workflow
n/a

## Bugfix / Desired workflow
n/a

## Questions
n/a

## Relevant issues
n/a

## Checklist
- [ ] linted with flake8?
- [ ] (if appropriate) unit tests added?

### CODE REVIEW

**Review by shihchengli — COMMENTED** [review]
Thanks for making this PR. There are two minor things you might consider adding to the notebook:
1. Include metrics within the model (e.g., `metric_list = [nn.metrics.RMSEMetric(), nn.metrics.MAEMetric()]`) in the notebook.
2. Add the ASHAScheduler to the notebook, as we use it by default.

**Comment by shihchengli** [comment]
The test fails because the filename has changed. Changing the filename of "best_params.json" to "best_config.json" in test_hyperopt_quick within tests/unit/utils/test_save_load_mol.py can fix it.

**Comment by kevingreenman** [comment]
should we add another notebook with an example of using hpopt with optuna, in addition to the hyperopt example? (after #783 is resolved)

File: chemprop/cli/hpopt.py [hunk 1]
```diff
@@ -242,6 +243,9 @@ def update_args_with_config(args: Namespace, config: dict) -> Namespace:
             case "init_lr_ratio":
                 setattr(args, "init_lr", value * args.max_lr)
 
+            case "batch_size" | "depth" | "ffn_num_layers" | "ffn_hidden_dim" | "message_hidden_dim":
```
[the remark this answers is not in the document]
Comment (hwpang) [reply chemprop/cli/hpopt.py:246]: This will be fixed in https://github.com/chemprop/chemprop/pull/873 and this line can be removed after that.

**Comment by shihchengli** [comment]
@hwpang here are my final comments about the notebook
1. replace the `uniform` with `qrandint` in the search space
2. Since the default scheduler is changed to `FIFO` in #896, please replace the `ASHAScheduler` with `FIFOScheduler`
3. As the optuna is supported via #888, instead of giving a new notebook about how to use optuna, you can add a comment in the file before the `HyperOptSearch` is used to generate the object of the search algorithm in the example, indicating that they can also use optuna by `search_alg = OptunaSearch()`, and change the notebook file name to `hpopting.ipynb`
4. remove the last empty cell
5. (Optional) you can consider adding [`ray.init()`](https://docs.ray.io/en/latest/ray-core/api/doc/ray.init.html#ray-init) before running ray

**Comment by hwpang** [comment]
@shihchengli Thanks for reviewing! I have updated according to your comments

**Comment by shihchengli** [comment]
There is a RayTaskError(OutOfMemoryError) in the notebook. I guess you are using estes cluster to test this notebook, and those two GPUs in the cluster are fully occupied now. You can try running this on the other cluster or use a CPU to test it, as we will also run the test in CI and we only have CPUs. I think it is fine to use the CPU in the notebook and point out the way to use the GPU.

File: chemprop/cli/hpopt.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (shihchengli) [reply chemprop/cli/hpopt.py:256]: Do we still need this after #873?

[the remark this answers is not in the document]
Comment (hwpang) [reply chemprop/cli/hpopt.py]: Good catch! We don't need this line anymore. Will drop it.

**Comment by shihchengli** [comment]
I noticed a minor thing that needed to be changed in the notebook. You can remove `find_unused_parameters=True` from the notebook as we fixed it via #883.

**Comment by shihchengli** [comment]
The error can be solved by changing the CI configuration to run test on hpopting.ipynb exclusively with Python 3.11.

**Comment by shihchengli** [comment]
You can do something like this, although it's not very elegant.
      
```
  - name: Test notebooks (excluding hpopting.ipynb)
    shell: bash -l {0}
    run: |
      python -m pip install matplotlib
      pytest --no-cov -v --nbmake $(find examples -name '*.ipynb' ! -name 'hpopting.ipynb')
  - name: Test hpopting.ipynb (Python 3.11 only)
    if: ${{ matrix.python-version == '3.11' }}
    shell: bash -l {0}
    run: |
      python -m pip install matplotlib
      pytest --no-cov -v --nbmake examples/hpopting.ipynb
```

**Comment by hwpang** [comment]
@shihchengli Thanks for the review and the suggestion! I have updated the CI accordingly

**Pull request merged**